### PR TITLE
LPS-94709 frontend-js-web - Modernize formatXML utility

### PR DESCRIPTION
**Pull request by wincent**
Pushing a fresh copy of https://github.com/wincent/liferay-portal/pull/26 here for review because GitHub makes viewing long-lived PRs unbearable.

### CODE REVIEW

**Comment by wincent** [comment]
ci:test:sf

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			</a:note>`;
+
+		expect(formatXML(receivedContent)).toEqual(expectedContent);
+	});
+
+	it('should return an XML string if content parameter is an XML string and options parameter is provided with custom values', () => {
+		const options = {newLine: 'line', tagIndent: 'tag'};
+
+		const expectedContent =
+			'line<?xml xlmns:a="http://www.w3.org/TR/html4/" version="1.0" encoding="UTF-8"?>line' +
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: That `line` thing caught me a bit off guard.  A more realistic use case would probably be `newLine: '\n'`. What do you think? Likewise, for `tagIndent` you could try something realistic, like, say, "  " (two spaces).

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			'tag<a:body>FooBarBaz!</a:body>line' +
+			'</a:note>';
+
+		const receivedContent = `<?xml xlmns:a="http://www.w3.org/TR/html4/" version="1.0" encoding="UTF-8"?>
+			<!DOCTYPE note>
+
+			<a:note>  					<a:to>Foo</a:to>
+				<a:from>Bar</a:from><a:heading>FooBar</a:heading>
+								<a:body>FooBarBaz!</a:body>
+			</a:note>`;
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Can we swap the order in these tests to make them more readable? ie. input before expected output.

```
const input = '...';

const expectedOutput = '...';

expect(...);
```

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+import formatXML from '../../../src/main/resources/META-INF/resources/liferay/util/format_xml.es';
+
+describe('Liferay.Util.formatXML', () => {
+	it('should throw error if content parameter is not a string', () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Hot off the press, we should probably write these `it()` without "should" as per [the guidelines](https://github.com/liferay/liferay-frontend-guidelines/blob/4ab3e533640366d2f7807569c1ba1985249f6b80/guidelines/general/testing.md#start-it-descriptions-with-a-verb-not-with-should). (Seeing as I wrote the guidelines, this is kind of like editing Wikipedia and then sending somebody a link to the article that you just edited.)

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+	it('should return an XML string if content parameter is an XML string and options parameter is not provided', () => {
+		const expectedContent =
+			'\r\n<?xml xlmns:a="http://www.w3.org/TR/html4/" version="1.0" encoding="UTF-8"?>\r\n' +
+			'<!DOCTYPE note>\r\n' +
+			'<a:note>\r\n' +
+			'	<a:to>Foo</a:to>\r\n' +
+			'	<a:from>Bar</a:from>\r\n' +
+			'	<a:heading>FooBar</a:heading>\r\n' +
+			'	<a:body>FooBarBaz!</a:body>\r\n' +
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Seeing as tabs are invisible, it might be helpful to put an explicit `\t` in the strings to make it clear what the expected whitespace is (seeing as this is essentially an assertion about whitespace).

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import formatXML from '../../../src/main/resources/META-INF/resources/liferay/util/format_xml.es';
+
+describe('Liferay.Util.formatXML', () => {
+	it('should throw error if content parameter is not a string', () => {
+		expect(() => formatXML({})).toThrow('must be a string');
+	});
+
+	it('should return an XML string if content parameter is an XML string and options parameter is not provided', () => {
+		const expectedContent =
+			'\r\n<?xml xlmns:a="http://www.w3.org/TR/html4/" version="1.0" encoding="UTF-8"?>\r\n' +
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Is the intent of the `slice()` call in the implementation to trim off the leading newline here? If so, does this mean it isn't working?

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js (hunk 1 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: This was supposed to be an edge case to test if the utility behaves as it's supposed to even if you add some silly parameters, but sure I can add `newLine: '`n' and `tagIndent: '  '`

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js (hunk 3 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Nice, should the file name be changed too, to abide by the laws of our guideline overlords as described here? https://github.com/liferay/liferay-frontend-guidelines/blob/4ab3e533640366d2f7807569c1ba1985249f6b80/guidelines/general/file_names.md

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js (hunk 4 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: I agree

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js (hunk 5 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Yes it is, it is removing it but only once

**Comment by kresimir-coko** [comment]
I force pushed the changes but unsure where to send them due to this PR not being linked to my branch.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js (hunk 3 above)
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/format_xml.es.js]: Excellent question, but the answer is "not yet". We'll do a mass change that renames all the files at once because Brian prefers things like that to happen atomically.

**Comment by wincent** [comment]
> I force pushed the changes but unsure where to send them due to this PR not being linked to my branch.

Thanks @kresimir-coko. Based on your updated branch I created a "Final version revisited part IV" here: https://github.com/wincent/liferay-portal/pull/36
